### ISSUE TITLE
Switching to queue time model

### ISSUE THREAD

**Issue by Coolthulhu** [CleverRaven/Cataclysm-DDA#20459]
Current time model has everything happen in waves:
* All creatures start their turns and gain moves
* All creatures move until their moves are `<1`
* All vehicles gain moves
* Until all vehicles have no moves, pick the vehicle with most moves and have it move
* All fields are processed

This is simple and usually good enough, but it may be worth considering switching to a model where everything happens as soon as possible, not when the category for this action comes up.

That is:
* Every entity that "deserves" own position in queue starts somewhere in the queue
* After executing a move, the entity is moved back in the queue by amount dependent on its current speed and move cost of the action
* After a creature moves, global time is advanced until the next entity in the queue is reached
* There are no "turns" for creature actions, turns only exist to have fields and effects tick down and act on synchronized global "clock"

Effects of this would be:
* Very fast creatures no longer multi-move each other, every creature "ends its turn" after a move
* Creatures and vehicles exist on the same time-scale and equal speed creatures and vehicles don't multi-move each other
* A very fast creature can have the reflexes to avoid a car charging at it in the very last moment rather than having to do it before the car is x tiles from it
* Vehicle driving gets a natural scaling effect: executing a vehicle action no longer moves the vehicle by entire "burst" of tiles, so having turning right cost 0.5 seconds would make perfect sense
* "Move synchronizing" tactics become less spoily: the tactic would be "wait until other creature moves" (with a specific keybind for that) rather than "manually count creature's moves to figure out a pattern in which the creature's turns and game turns intersect"

This would be an alternative to decreasing time per turn. Changing turn time would require changing the same spots as changing a time model, but would not be as future-proof as a new time model.

**Comment by kevingranade** [comment]
This is fairly nasty for both code complexity and (theoretical) memory locality.  It forces there to be some central action queue that dispatches to any given actor, this makes it harder to add caching that's pertinent to any given kind of operation.

AFAICT all of the pros of this would be accomplished by the much simpler alternative of making turns happen faster.
I'm not sure we want vehicle control to be continuous like that, it requires about 5x as many keypresses for the same distance travelled.
The final point doesn't even require this or shorter turns, just the keybind you mention.

Can you elaborate on how this kind of interleaved turn sequence would be more future-proof?

**Comment by Coolthulhu** [comment]
>  It forces there to be some central action queue that dispatches to any given actor, this makes it harder to add caching that's pertinent to any given kind of operation.

You mean caches that currently exist at per-turn basis?
Yeah, those would need to be rewritten to be less "dirty all cache" and more "dirty those x tiles of cache".

> AFAICT all of the pros of this would be accomplished by the much simpler alternative of making turns happen faster.

It wouldn't be much simpler once we included performance into the equation. At this point the difference in complexity between discrete turns and queue would blur heavily.
For example, if we switched to 1 second granularity, we'd have to rewrite many functions to use arbitrary time or stuff like vehicle processing and field processing would eat up a ton of time.

> I'm not sure we want vehicle control to be continuous like that, it requires about 5x as many keypresses for the same distance travelled.

Unless there was a keybind for a longer pause. Or a different driving model.

> The final point doesn't even require this or shorter turns, just the keybind you mention.

No, it isn't possible with current time model. At least without implementing weird out-of-turn turns where creature's movement triggers player's turn. Which would actually be a good setup for queue model.

> Can you elaborate on how this kind of interleaved turn sequence would be more future-proof?

It would solve the problem of turn syncing once and for all, in a way that arbitrary turns can only solve for specific sets of speed.

Consider a monster with speed that isn't 100, 50 or 25. This monster will act in a pattern: once every ~n turns, it will execute one action more (or less) than in other turns.
The player wants to act on such turns, that the monster will have the least number of turns afterwards.
For example, hulk has speed 130, so it's "turn pattern" is `[1, 1, 1, 2, 1, 1, 2, 1, 1, 2]` - that is, it gets two turns every ~3 turns. So the player wants to bait the hulk into doing the double turn, so that the next turn it does only one turn.

While higher granularity pushes the problem further back (decreases impact), arbitrary time as in the queue model solves it for good.

Plus, queue model is just elegant: everything happens when it happens, not during the moment that started before it happened and didn't end yet.
